### PR TITLE
Fix the iou request for "unknown" users

### PR DESCRIPTION
**Pull request by ShogunFire**
### Details

Due to the privacy changes, the login is null for users that haven't sent message to us so when we made an iou request the payerEmail sent to the api was null which was creating an error. Since we are actively creating a request or creating a chat with this user using his email/phone, we can use that as the payerEmail when there is no login

### Fixed Issues

$ https://github.com/Expensify/App/issues/22884
PROPOSAL: https://github.com/Expensify/App/issues/22884#issuecomment-1642406559

### Tests

1. Create a chat with a user that you have never sent a message to
2. In this chat, click on + and make an iou request 
3. Send some message.
4. Request money again.
5. Verify that there is only one IOU Preview and that it is the last message, also verify that there are two replies below this preview

- [x] Verify that no errors appear in the JS console

### Offline tests

1. While offline create a chat with a user that you have never sent a message to
2. In this chat, click on + and make an iou request 
3. Send some message.
4. Request money again.
5. Verify that there is only one IOU Preview and that it is the last message
6. Come back online and verify that the iou preview now has 2 replies

### QA Steps

1. Create a chat with a user that you have never sent a message to
2. In this chat, click on + and make an iou request 
3. Send some message.
4. Request money again.
5. Verify that there is only one IOU Preview and that it is the last message, also verify that there are two replies below this preview

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/19537677/1ec4ccd3-a9b8-4e8d-a619-fda18e892650

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/19537677/fa12f7e8-958f-45c6-a53d-a10bb410abb3

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/19537677/871b79a4-61cc-4598-ad9a-a6ab20449ca0

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/19537677/2dbd5f11-287a-44de-847f-84330cd29b12

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/19537677/7160de57-3dd8-4f1c-85a2-c048c2f5e03b

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/19537677/464882f0-cf09-4d73-88a6-ac637a759fe9

</details>

### CODE REVIEW

**Comment by ShogunFire** [comment]
I think it's already working for split which for some reasons still has the email in the logins, buildOptimisticIOUReport is also used for sendMoney but if I am not mistaken this functionality is disabled.

**Comment by ShogunFire** [comment]
PR ready for review minus the android video (I have some troubles building)

File: src/libs/actions/IOU.js [hunk 1]
```diff
@@ -315,7 +315,7 @@ function buildOnyxDataForMoneyRequest(
  *
  */
 function requestMoney(report, amount, currency, payeeEmail, payeeAccountID, participant, comment, receipt = undefined) {
-    const payerEmail = OptionsListUtils.addSMSDomainIfPhoneNumber(participant.login);
+    const payerEmail = OptionsListUtils.addSMSDomainIfPhoneNumber(participant.login || participant.text);
```
[the remark this answers is not in the document]
Comment (ShogunFire) [reply src/libs/actions/IOU.js:318]: I agree with you, I had sent this on slack for another issue: https://expensify.slack.com/archives/C01GTK53T8Q/p1690214768266839 

I was saying if we create the chat with another user, we know his login, I think that's the same for a iou request. Someone answered me that this weird behaviour was intended so I don't know...

[the remark this answers is not in the document]
Comment (jasperhuangg) [reply src/libs/actions/IOU.js:318]: What is the default email that you are referring to?

[the remark this answers is not in the document]
Comment (ShogunFire) [reply src/libs/actions/IOU.js:318]: So what do we decide ?

File: src/libs/ReportUtils.js [hunk 2]
```diff
@@ -1653,13 +1653,13 @@ function buildOptimisticTaskCommentReportAction(taskReportID, taskTitle, taskAss
...
- *
+ * @param {String} unknownPayerEmail - The mail or phone of the payer, usefull when we don't know the login of the user
  * @returns {Object}
  */
-function buildOptimisticIOUReport(payeeAccountID, payerAccountID, total, chatReportID, currency, isSendingMoney = false) {
+function buildOptimisticIOUReport(payeeAccountID, payerAccountID, total, chatReportID, currency, isSendingMoney = false, unknownPayerEmail = '') {
     const formattedTotal = CurrencyUtils.convertToDisplayString(total, currency);
     const personalDetails = getPersonalDetailsForAccountID(payerAccountID);
-    const payerEmail = personalDetails.login;
+    const payerEmail = personalDetails.login || unknownPayerEmail;
```
[the remark this answers is not in the document]
Comment (ShogunFire) [reply src/libs/ReportUtils.js:1662]: Honestly I don't see any alternative if we keep the privacy change as it is

File: src/libs/actions/IOU.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShogunFire) [reply src/libs/actions/IOU.js:318]: Honestly I don't see any alternative if we keep the privacy change as it is

[the remark this answers is not in the document]
Comment (ShogunFire) [reply src/libs/actions/IOU.js:318]: Or maybe we can do this change in the getOptions so that it is made for every search

**Comment by mountiny** [comment]
Yeah I can see why this does not look ideal, I think we however need to make a change like this right now and document it better to make it clear why we are doing this. It might be temporary but worth it.

File: src/libs/ReportUtils.js [hunk 3]
```diff
@@ -1653,13 +1653,13 @@ function buildOptimisticTaskCommentReportAction(taskReportID, taskTitle, taskAss
  * @param {String} chatReportID - Report ID of the chat where the IOU is.
  * @param {String} currency - IOU currency.
  * @param {Boolean} isSendingMoney - If we send money the IOU should be created as settled
- *
+ * @param {String} unknownPayerEmail - The mail or phone of the payer, usefull when we don't know the login of the user
  * @returns {Object}
  */
-function buildOptimisticIOUReport(payeeAccountID, payerAccountID, total, chatReportID, currency, isSendingMoney = false) {
+function buildOptimisticIOUReport(payeeAccountID, payerAccountID, total, chatReportID, currency, isSendingMoney = false, unknownPayerEmail = '') {
```
Comment (mountiny) [inline src/libs/ReportUtils.js:1659]: Lets just call it payerEmail, but in the docs explain the situation
```suggestion
function buildOptimisticIOUReport(payeeAccountID, payerAccountID, total, chatReportID, currency, isSendingMoney = false, payerEmail = '') {
```

**Review by mountiny — COMMENTED** [review]
@ShogunFire @parasharrajat we also need the same fix for Split bill right?

**Comment by ShogunFire** [comment]
> @ShogunFire @parasharrajat we also need the same fix for Split bill right?

I think yes. Actually I wanted to reverify this but there seems to be another error right now when we split bill with unknown users due to formating "user1 and user 2" when they don't have login here: https://github.com/Expensify/App/blob/e3074c49a7938b9d978b48721fdbadd8980bf644/src/libs/actions/IOU.js#L615-L617  

Anyways I will Make the change because yes I think it's necessary

**Comment by mountiny** [comment]
Yep I think thats highly possible and we should try to make it safely accessed so it looks fine at the same time

**Comment by ShogunFire** [comment]
I made the changes, and resolved the conflict but it's still showing, what can I do to resolve it ?

File: src/libs/ReportUtils.js (hunk 3 above)
Comment (ShogunFire) [reply src/libs/ReportUtils.js]: There is already a payerEmail variable in the function, checking if we need it now

**Comment by ShogunFire** [comment]
Ok I made some changes, I made payerEmail required parameter and removed the one that was in the function, tell me what you think. 

Also I am not sure why this PR still thinks there is conflict, I think I resolved them already

**Comment by ShogunFire** [comment]
And I don't think this is in the scope of this issue but when I am offline this amount is not updated nor the number of replies (The number of requests is updated)
image

**Comment by ShogunFire** [comment]
@dukenv0307 has a good point here: https://github.com/Expensify/App/issues/23038#issuecomment-1687332985 That method doesn't work if the user changed his display name.

I have another sketchy method that seem to work for all users, 
When the user change his display name the only place where we can find his email seem to be in `report.participants`, unfortunely this is just a list of all emails of the report without any way to search a specific user, but the order of report.participants seem to be the same than the order of report.participantAccountIDs. So in navigateToNextPage here  https://github.com/Expensify/App/blob/5a373bf656d4c1f8c16d037eeecd4a5a715d206d/src/libs/actions/IOU.js#L1789-L1794, we can set the login of the participants like this:

```
var participants = [];

           
for (let i = 0; i < report.participantAccountIDs.length; i++) {
  const accountID = report.participantAccountIDs[i];
  if (currentUserAccountID !== accountID){
      participants.push({accountID, selected: true, login: report.participants[i]})
  }
}
```    

If we go with that method, the other changes are unnecessary

**Comment by ShogunFire** [comment]
Otherwise I recommend this but it would take way more effort and time:

"I have  a proposition about the privacy change that is causing so many issues. Why don't we create another field email or privateLogin that we will never display and if needed we can even encrypted it so that it doesn't appear in the network. And use that field to do all the things that we are currently doing with the login field"

**Comment by mountiny** [comment]
Argh I hate this thing, I think we discussed this couple of times but it was deprioritized. 

could we maybe just Blcok requesting money from chat with user where we dont have email, ie you can request from Global create from anyone but you cannot keep requesting until that perosn responds to you within the chat

**Comment by ShogunFire** [comment]
I made the change

**Comment by ShogunFire** [comment]
There seems to have an issue, when we create a chat with a unknown user the money request option is showing for a few seconds (probably until the server response) I will try to fix that

**Comment by ShogunFire** [comment]
:( so in getOptions we add the login to the userToInviteOption, I don't think it is safe to remove that
https://github.com/Expensify/App/blob/9a0dd42f675b02ea0a0bcafabde28f5f7cf03c97/src/libs/OptionsListUtils.js#L833

So maybe we can also prevent money request if the report is optimistic, is there a reason we don't add isOptimisticReport: true in this method: https://github.com/Expensify/App/blob/9a0dd42f675b02ea0a0bcafabde28f5f7cf03c97/src/libs/ReportUtils.js#L2199

**Comment by mountiny** [comment]
@ShogunFire thanks for trying

I think we can hold on this one until we have a clearer plan as it does not seem as ultimately the best direction to head into.

**Comment by ShogunFire** [comment]
Can I try one last thing ? I can start implementing this

> Why don't we create another field email or privateLogin that we will never display and if needed we can even encrypted it so that it doesn't appear in the network. And use that field to do all the things that we are currently doing with the login field

I don't think I need back end changes, I can just fill the new field optimistically  with the login here:

https://github.com/Expensify/App/blob/867d61e237e6a4e0affaf71e6735d0cdf6969a70/src/libs/actions/Report.js#L496-L502

Then for the payerEmail I can use that field instead of login or text here: https://github.com/Expensify/App/blob/40f354b36905d97513ccc52add7859cfd9c6aa9e/src/libs/actions/IOU.js#L308

**Comment by mountiny** [comment]
the best solution so far is to update those flows to work with `accountIDs` as we have those, I think any other proposed solution so far is quite "dirty"

**Comment by ShogunFire** [comment]
The accountIDs created optimistically and returned by the backend can be different right ? Then I don't know how making iou request offline based on something that can change would work. Anyways, seems like there is not much I can do right now if we go with that option

**Comment by mountiny** [comment]
@ShogunFire you are right, that is one of the challenges there

**Comment by mountiny** [comment]
This is the issue https://github.com/Expensify/App/issues/22480

**Comment by mountiny** [comment]
> The accountIDs created optimistically and returned by the backend can be different right ? Then I don't know how making iou request offline based on something that can change would work. Anyways, seems like there is not much I can do right now if we go with that option

I think we would have to use both email and accountID. AccountID would be a fallback, when you are creating this offline you know the email so you can use that, then once you sign  out and sign back in you had to be online and you had to get the correct accountID

**Comment by ShogunFire** [comment]
Honestly it doesn't seem like a pretty solution. I feel like the privacy change should have just been a display issue. Not displaying the login, but still using it. 

I would personally prefer to have an email field that we would never display or a boolean shouldDisplayLogin field. It would require way less changes and seems like an easier solution.

**Comment by mountiny** [comment]
> Not displaying the login, but still using it.

this would not prevent people from just scraping the contact details from the public chat rooms, we cannot return this data from BE

**Comment by mountiny** [comment]
I think that was one of the options when the project was planned but it was not chosen for reasons I am not familiar with

**Pull request closed without merging**